flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/TableStreamingReader.java

```diff
@@ -82,31 +85,19 @@ public TableStreamingReader(
     }
 
     @Nullable
-    public Iterator<RowData> nextBatch() throws IOException {
+    public Iterator<RowData> nextBatch() throws Exception {
         if (enumerator == null) {
             DataTableScan scan = table.newScan();
             if (predicate != null) {
                 scan.withFilter(predicate);
             }
-            DataTableScan.DataFilePlan plan = scan.plan();
-            if (plan.snapshotId == null) {
-                return null;
-            }
-            long snapshotId = plan.snapshotId;
             enumerator =
-                    new SnapshotEnumerator(
-                            table.location(),
-                            scan.withIncremental(true),
-                            table.options().changelogProducer(),
-                            snapshotId);
-            return read(plan);
-        } else {
-            SnapshotEnumerator.EnumeratorResult result = enumerator.call();
-            if (result == null) {
-                return null;
-            }
-            return read(result.plan);
+                    new DeltaSnapshotEnumerator(
```

Comment:
Maybe we can put these in constructor?
The change here is that initialization does not need to generate a batch.